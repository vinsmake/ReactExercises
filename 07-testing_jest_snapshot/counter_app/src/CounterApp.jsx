import {useState} from 'react';
import PropTypes from 'prop-types';

export function CounterApp ({defaultValue}) {

    //this is the hook. 
    //--variable, action
    //[counting, setCounting]
    // hook to use (initial valor)
    //= useState( 0 );
    //useState changes the state of the --variable
    const [counting, setCounting] = useState( defaultValue );

    function plusClick() {
        //this is important because es changing the state, it's not counting++, its setCounting = counting +1; 
        setCounting(counting + 1)
    };

    const substractClick = () => {
        setCounting(counting - 1)
    };

    function resetClick(params) {
        setCounting(defaultValue)
    }


    return(
    <>
        <h2>This is a Counter App</h2>
        <h4>Which counts +1 every time it's clicked</h4>
        <h3 data-testid="IDcounterApp">{counting}</h3>
        <button onClick={substractClick}>
            -1</button>
        <button onClick={resetClick}>
            Reset</button>
        <button onClick={plusClick}>
            +1</button>
    </>
    );
}

CounterApp.propTypes = {
    //this means that counter mus be a number, this PropTypes comes from the import PropTypes. isRequired means that that valor must be defined
    counting: PropTypes.number.isRequired
}

CounterApp.defaultProps = {

}

