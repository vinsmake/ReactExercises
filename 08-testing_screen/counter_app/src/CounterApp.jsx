import {useState} from 'react';
import PropTypes from 'prop-types';

export function CounterApp ({defaultValue}) {

    //this is the hook. 
    //--variable, action
    //[counting, setCounting]
    // hook to use (initial valor)
    //= useState( 0 );
    //useState changes the state of the --variable
    const [counting, setCounting] = useState( defaultValue );

    function plusClick() {
        //this is important because es changing the state, it's not counting++, its setCounting = counting +1; 
        setCounting(counting + 1)
    };

    const substractClick = () => {
        setCounting(counting - 1)
    };

    function resetClick(params) {
        setCounting(defaultValue)
    }


    return(
    <>
        <h2>This is a Counter App</h2>
        <h4>Which counts +1 every time it's clicked</h4>
        <h3 data-testid="IDcounterApp" id="IDcounterApp">{counting}</h3>
        <button data-testid="test-btn--1" onClick={substractClick}>
            -1</button>
        <button data-testid="test-btn-reset" aria-label='btn-reset' onClick={resetClick}>
            Reset</button>
        <button data-testid="test-btn-+1" onClick={plusClick}>
            +1</button>
    </>
    );
}

CounterApp.propTypes = {
    //this means that counter mus be a number, this PropTypes comes from the import PropTypes. isRequired means that that valor must be defined
    counting: PropTypes.number
}

CounterApp.defaultProps = {

}

