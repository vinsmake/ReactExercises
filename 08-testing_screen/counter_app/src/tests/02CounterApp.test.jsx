import { fireEvent, render, screen } from '@testing-library/react';
import { CounterApp } from '../CounterApp';

describe('Testing CounterApp', () => { 
    const defaultValue = 0;
    test('should match with the snapshot', () => { 

        const { container } = render( <CounterApp defaultValue = {defaultValue} /> )
        expect( container ).toMatchSnapshot();

     })

     test('should exist', () => { 
        
        render( <CounterApp defaultValue = {defaultValue} /> )
        
        expect( screen.getByTestId("IDcounterApp")).toBeTruthy();
      })

      test('should be 0', () => { 
        
        render( <CounterApp defaultValue = {defaultValue} /> )
        
        expect( screen.getByTestId("IDcounterApp").innerHTML).toBe("0");
      })

      test('should be 0 by rol heading lv3', () => { 
        
        render( <CounterApp defaultValue = {defaultValue} /> )
        
        expect( screen.getByRole('heading', { level: 3 }).innerHTML ).toContain("0");
      })

      test('should increase when clic', () => { 
        
        render( <CounterApp defaultValue = {defaultValue} /> );
        fireEvent.click( screen.getByTestId("test-btn-+1") )
          expect( screen.getByTestId("IDcounterApp").innerHTML).toBe("1");
       })


      test('should decrease when clic', () => { 
        
        render( <CounterApp defaultValue = {defaultValue} /> );
        fireEvent.click( screen.getByTestId("test-btn--1") )
          expect( screen.getByTestId("IDcounterApp").innerHTML).toBe("-1");
       })


      test('should reset when clic', () => { 
        
        render( <CounterApp defaultValue = {defaultValue} /> );
        fireEvent.click( screen.getByTestId("test-btn-reset") )
          expect( screen.getByTestId("IDcounterApp").innerHTML).toBe("0");
       })


      test('should function every button when clic', () => { 
        //sujeto de prueba
        render( <CounterApp defaultValue = {defaultValue} /> ); 

        //estimulos
        fireEvent.click( screen.getByTestId("test-btn-reset") )
        fireEvent.click( screen.getByTestId("test-btn--1") )
        fireEvent.click( screen.getByTestId("test-btn--1") )
        fireEvent.click( screen.getByTestId("test-btn-reset") )
        fireEvent.click( screen.getByTestId("test-btn-+1") )
        fireEvent.click( screen.getByTestId("test-btn-+1") )
        fireEvent.click( screen.getByRole('button', {name: 'btn-reset'}) )

        //resultados
          expect( screen.getByTestId("IDcounterApp").innerHTML).toBe("0");
       })

       

 })