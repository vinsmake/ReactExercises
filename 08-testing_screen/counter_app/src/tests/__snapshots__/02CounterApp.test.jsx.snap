// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CounterApp should match with the snapshot 1`] = `
<div>
  <h2>
    This is a Counter App
  </h2>
  <h4>
    Which counts +1 every time it's clicked
  </h4>
  <h3
    data-testid="IDcounterApp"
    id="IDcounterApp"
  >
    0
  </h3>
  <button
    data-testid="test-btn--1"
  >
    -1
  </button>
  <button
    aria-label="btn-reset"
    data-testid="test-btn-reset"
  >
    Reset
  </button>
  <button
    data-testid="test-btn-+1"
  >
    +1
  </button>
</div>
`;
